### PR TITLE
fix(collab): validate and rate-limit comment mutations

### PR DESCRIPTION
**Pull request by Ayush7614**
## Summary
- Validate comment/reply shapes in the collab Durable Object before appending them to the shared snapshot.
- Cap body length like chat (2000) and throttle mutations (250ms) so peers cannot inflate or poison project comments.
- Broadcast only the sanitized action; return a `bad-message` error to the sender on invalid payloads.

## Test plan
- [x] `node --import tsx --test tests/collab-comment-validate.test.ts tests/collab-protocol.test.ts`
- [x] In a live collab session, add a normal comment and confirm peers receive it
- [x] Attempt an oversized/malformed mutation and confirm it is rejected without updating peers

## Summary by CodeRabbit

* **New Features**
  * Added validation and sanitization for comment creation, replies, resolution changes, and deletion.
  * Added safeguards for comment content, authors, anchors, timestamps, and nested replies.
  * Added per-connection rate limiting for comment updates.

* **Bug Fixes**
  * Prevented malformed or unsupported comment data from being stored or broadcast.
  * Excluded invalid replies while preserving valid comments.
  * Prevented duplicate comments and replies.
  * Ensured valid updates continue to be broadcast when saving fails.

* **Tests**
  * Added comprehensive coverage for validation, sanitization, edge cases, and invalid inputs.

### CODE REVIEW

File: workers/collab/src/session.ts [hunk 1]
```diff
@@ -671,53 +674,125 @@ export class CollabSession extends DurableObject<Env> {
...
-        // Ignore snapshot mutation update errors defensively
+        // Snapshot mutation failed; still fan out the validated action below so
+        // peers stay in sync (the next full snapshot will reconcile storage).
       }
     }
 
-    // Exclude the sender (ws) so they don't receive their own mutation back.
-    // The sender already applied the change locally before calling sendCommentMutation.
-    this.broadcast(message, ws);
+    this.broadcast(sanitizedMessage, ws);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply workers/collab/src/session.ts:895]: Done in 0d753135 — the first half (persisting when no snapshot exists yet) landed in 803d8a9e; this commit closes the second half. New pure helper `preserveStoredComments` merges the stored `comments` list into an incoming full-project snapshot that omits the key (`serializeProject` drops it when a peer holds none), so a peer that hasn't merged the comment-mutation broadcasts can no longer clobber them. A project that supplies its own `comments` still wins, so an explicit delete-to-empty is never resurrected, and the merged project is what gets broadcast, which heals the drifted sender too.

File: workers/collab/src/comment-validate.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  if (!anchor) return null;
+
+  const author = validateAuthor(o.author);
+  if (!author) return null;
+
+  if (typeof o.body !== "string") return null;
+  const body = o.body.slice(0, MAX_COMMENT_BODY_LENGTH);
+  if (!body.trim()) return null;
+
+  const createdAt = typeof o.createdAt === "string" ? o.createdAt : new Date().toISOString();
```
[the remark this answers is not in the document]
Comment (giswqs) [reply workers/collab/src/comment-validate.ts]: Already fixed in 52a926c3 — `validateComment` and `validateReply` both gate `createdAt` on `!Number.isNaN(Date.parse(o.createdAt))` and fall back to `new Date().toISOString()`, exactly as suggested (comment-validate.ts:125-128 and :171-174). Covered by the "falls back to now for unparseable createdAt" test.

File: workers/collab/src/comment-validate.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    if (!Array.isArray(o.lngLat) || o.lngLat.length !== 2) return null;
+    const [lng, lat] = o.lngLat;
+    if (!finite(lng) || !finite(lat)) return null;
+    return { type: "point", lngLat: [lng, lat] };
+  }
+
+  if (o.type === "feature") {
+    if (typeof o.layerId !== "string" || !o.layerId) return null;
+    if (typeof o.featureId !== "string" && typeof o.featureId !== "number") return null;
+    if (typeof o.featureId === "string" && !o.featureId) return null;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply workers/collab/src/comment-validate.ts]: Already fixed in 803d8a9e — a numeric `featureId` now goes through the same `finite()` guard as the `lngLat` coordinates (`if (typeof o.featureId === "number" && !finite(o.featureId)) return null;`, comment-validate.ts:62), so NaN/Infinity are rejected rather than silently orphaning the anchor.

File: workers/collab/src/session.ts [hunk 4]
```diff
@@ -671,53 +674,125 @@ export class CollabSession extends DurableObject<Env> {
...
         let updatedComments = comments;
 
-        if (action.type === "add") {
-          if (!action.comment || typeof action.comment !== "object") return;
-          updatedComments = [...comments, action.comment as Record<string, unknown>];
-        } else if (action.type === "reply") {
-          if (!action.reply || typeof action.reply !== "object") return;
-          const replyObj = action.reply as Record<string, unknown>;
+        if (sanitizedAction.type === "add") {
+          updatedComments = [...comments, sanitizedAction.comment as Record<string, unknown>];
```
[the remark this answers is not in the document]
Comment (giswqs) [reply workers/collab/src/session.ts]: Done in 0d753135 — all three parts. (1) Total growth is now bounded by `MAX_COMMENTS_PER_SESSION` (500), mirroring `CHAT_HISTORY_LIMIT`; an `add` past the cap gets a `bad-message` error instead of appending. (2) The serialized snapshot is checked against `MAX_SNAPSHOT_BYTES` before the `storage.put`, matching `handleSnapshot`. (3) The divergence you described is gone: a failed persist no longer falls through to `this.broadcast` — the sender gets an error frame and the fan-out is skipped, so peers never hold a comment that storage doesn't have. I kept a hard cap rather than chat-style eviction because silently evicting someone else's review comment is worse than refusing a new one.

File: workers/collab/src/comment-validate.ts [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  return {
+    id: o.id,
+    anchor,
+    author,
+    body,
+    createdAt,
+    resolved: Boolean(o.resolved),
+    replies,
+  };
+}
```
[the remark this answers is not in the document]
Comment (giswqs) [reply workers/collab/src/comment-validate.ts:173]: Already fixed in 803d8a9e — `MAX_ID_LENGTH` (200) plus an `isBoundedId()` helper now gate `comment.id`, `reply.id`, `anchor.layerId`, and a string `anchor.featureId` (comment-validate.ts:12-23, :59-61, :113, :162). `session.ts` uses the same helper for the `commentId` on reply/toggle-resolve/delete, so an unbounded id can't reach storage through any action. Covered by the "rejects an oversized id" tests.

File: workers/collab/src/session.ts [hunk 6]
```diff
@@ -671,53 +757,86 @@ export class CollabSession extends DurableObject<Env> {
...
           updatedComments = comments.map((c) => {
-            if (!c || typeof c !== "object" || c.id !== action.commentId) return c;
+            if (!c || typeof c !== "object" || c.id !== sanitizedAction.commentId) return c;
             const existingReplies = Array.isArray(c.replies)
               ? (c.replies as Record<string, unknown>[])
               : [];
             if (existingReplies.some((r) => r && typeof r === "object" && r.id === replyObj.id))
               return c;
             return { ...c, replies: [...existingReplies, replyObj] };
           });
```
[the remark this answers is not in the document]
Comment (giswqs) [reply workers/collab/src/session.ts]: Already fixed in 803d8a9e — `handleCommentMutation` now returns a `bad-message` ("Invalid reply target.") when `comments.find(...)` yields no target, before the reply-limit check, so the no-op map and the broadcast are both skipped (session.ts:808-815). Matches the proposed structure.

File: workers/collab/src/session.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (giswqs) [reply workers/collab/src/session.ts:895]: Confirming it's intentional, and leaving this thread open for a maintainer to weigh in rather than resolving it myself.

The truncation/coercion is deliberate and does mirror the chat path: the relay must not fan out an unbounded string, and the sender is the one client that already has the content, so echoing a corrected copy back would need the client store to *replace* on echo (`addComment`/`replyToComment` currently de-dupe by id and ignore it) — a client-side change beyond the scope of this hardening PR.

Two things narrow the practical window since you flagged it:
- The resync you describe is now less likely to be lossy in the other direction — as of 0d753135 `handleSnapshot` preserves the stored `comments` when an incoming project omits them, so a host resync no longer silently drops comments it hasn't merged.
- Drift can only originate from a >2000-char body (or a name over 120 chars), which the normal UI path doesn't produce on its own; the comment textareas have no `maxLength`, so a paste can still exceed it.

If we want to close it properly, the clean fix is a client-side `maxLength` on the comment/reply textareas backed by a shared constant, so the server cap is never the first place a user learns about the limit. Happy to do that as a follow-up if you'd like it in this PR.
